kloud/provider: add Azure support

rjeczalik opened the pull request:
~~deps: https://github.com/koding/koding/pull/9182~~

rjeczalik committed:
provider/azure: initial import

rjeczalik committed:
provider/azure: rework to new stack API

rjeczalik committed:
provider/azure: fix bootstrap template codegen

rjeczalik commented:
ping

cihangir commented:
could you separate the sdk from this PR into another one? github ui does not show the diff

rjeczalik commented:
@cihangir Yup sorry, actually I forgot to split it - here it is: https://github.com/koding/koding/pull/9309

rjeczalik commented:
split and rebased

On 17 October 2016 at 19:06, Cihangir <notifications@github.com> wrote:

> could you separate the sdk from this PR into another one? github ui does
> not show the diff
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/koding/koding/pull/9241#issuecomment-254269473>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/ABG7IYLkhhp734Xn6EW3wV8eJzu0vi9Fks5q06segaJpZM4KSyba>
> .
>

cihangir commented on go/src/koding/kites/kloud/provider/azure/azure.go:
```diff
@@ -0,0 +1,207 @@
+}
+
+type PublishData struct {
+	Subscriptions []Subscription `xml:"PublishProfile>Subscription"`
```
xml :/

cihangir commented on go/src/koding/kites/kloud/provider/azure/azure.go:
```diff
@@ -0,0 +1,207 @@
+		return nil
+	}
+
+	var found bool
```
you could simplify with

```
  for _, sub := range pb.Subscriptions {
      if sub.ID == meta.SubscriptionID {
          return nil
      }
  }

  return errors.New("specified subscription ID does not exist")
```

rjeczalik replied:
yup, will do - when I coded it originally I had two more checks after pb.Subscriptions that's why no early return here

cihangir commented on go/src/koding/kites/kloud/provider/azure/azure.go:
```diff
@@ -0,0 +1,207 @@
+	vm "github.com/Azure/azure-sdk-for-go/management/virtualmachine"
+)
+
+var p = &provider.Provider{
```
you dont need to have this in global scope AFAIS

cihangir commented on go/src/koding/kites/kloud/provider/azure/bootstrap_test.go:
```diff
@@ -0,0 +1,28 @@
+)
+
+func TestBootstrapJSONTmpl(t *testing.T) {
+	cases := []*azure.BootstrapConfig{
```
could you please add test cases for other variables in this struct as well?

rjeczalik replied:
Done.

cihangir commented on go/src/koding/kites/kloud/provider/azure/bootstrap_test.go:
```diff
@@ -0,0 +1,28 @@
+
+		var v map[string]interface{}
+
+		if err = json.Unmarshal(tmpl, &v); err != nil {
```
you can have expected results in a table and check them here

rjeczalik replied:
I don't test for expected values, only for valid JSON - so for every case throwaway `v` variable is used.

cihangir commented on go/src/koding/kites/kloud/provider/azure/machine.go:
```diff
@@ -0,0 +1,100 @@
+}
+
+func (m *Machine) Start(ctx context.Context) (interface{}, error) {
+	id, err := m.AzureVMClient.StartRole(m.Meta().HostedServiceID, m.Meta().InstanceID, m.Meta().InstanceID)
```
could we do something like for all the functions?
```
	ctx, cancel := context.WithTimeout(ctx, 5*time.Minute)
	defer cancel()

	closeChan := make(chan struct{})
	go func() {
		<-ctx.Done()
		close(closeChan)
	}()
	WaitForOperation(closeChan)
```

cihangir commented on go/src/koding/kites/kloud/provider/azure/machine.go:
```diff
@@ -0,0 +1,100 @@
+	// for both in case this would get fixed.
+	e, ok := err.(*management.AzureError)
+	if !ok {
+		ee, ok := err.(management.AzureError)
```
srsly? sometimes with pointer sometimes not?

rjeczalik replied:
Currently they use struct values - without pointers, I'm handling also pointers if they had a second thought and fixed this eventually.

cihangir commented on go/src/koding/kites/kloud/provider/azure/stack.go:
```diff
@@ -0,0 +1,322 @@
+
+var tmpl = template.Must(template.New("").Parse(mustAsset("bootstrap.json.tmpl")))
+
+type BootstrapConfig struct {
```
linting issues

cihangir commented on go/src/koding/kites/kloud/provider/azure/stack.go:
```diff
@@ -0,0 +1,322 @@
+	cfg := &BootstrapConfig{
+		TeamSlug:           s.BootstrapArg().GroupName,
+		HostedServiceName:  "koding-hs-" + c.Identifier,
+		StorageServiceName: strings.ToLower("kodings" + c.Identifier),
```
s/kodings/koding-s- ?

rjeczalik replied:
Storage name can't have `-` characters, that's why it differs.

cihangir commented on go/src/koding/kites/kloud/provider/azure/stack.go:
```diff
@@ -0,0 +1,322 @@
+	}, nil
+}
+
+func (s *Stack) ApplyTemplate(c *stack.Credential) (*stack.Template, error) {
```
could you please separate this function into smaller ones?

cihangir replied:
could you write test?

cihangir commented on go/src/koding/kites/kloud/provider/azure/stack_test.go:
```diff
@@ -0,0 +1,73 @@
+      ManagementCertificate="..." />
+    <Subscription
+      ServiceManagementUrl="https://management.core.windows.net"
+      Id="8a523531-2d66-469d-9228-718ad2a58eff"
```
where these ids come from? safe to be public?

rjeczalik replied:
Yes, safe, they are not valid. If they are, then not ours ;-)

rjeczalik committed:
provider/azure: simplify (*Cred).Valid method

rjeczalik committed:
provider/azure: make provider a local variable

rjeczalik committed:
provider/azure: add inline documentation

rjeczalik committed:
provider/azure: move Storage from Bootstrap to Credential

rjeczalik committed:
provider/azure: more inline documentation

rjeczalik committed:
provider/azure: split ApplyTemplate into smaller methods

rjeczalik committed:
provider/azure: add test for ApplyTemplate

rjeczalik commented:
@cihangir All addressed, ptal.

ppknap commented on go/src/koding/kites/kloud/provider/azure/stack.go:
```diff
@@ -0,0 +1,351 @@
+}
+
+func mustAsset(s string) string {
+	p, err := Asset(s)
```
You can use `MustAsset` generated by go-bindata.

rjeczalik committed:
provider/azure: remove mustAsset

cihangir reviewed (approved):
we can merge after deployment

rjeczalik committed:
provider/azure: apply workaround for hcl.DecodeObject

rjeczalik committed:
provider: add Slice type

This works around the problem described in #9346 by using
map as a underlying representation of slice. This way
no slices are merged together since map's keys are unique.

When marshalled to JSON, the map is converted back to
an actual slice.

Fixes #9346.

rjeczalik commented on go/src/koding/kites/kloud/provider/azure/stack.go:
```diff
@@ -250,7 +254,7 @@ func (s *Stack) injectEndpointRules(vm map[string]interface{}) {
 	}
 
-	vm["endpoint"] = endpoints
+	vm["endpoint"] = provider.ToSlice(endpoints)
```
/cc @ppknap 

```
--- PASS: TestAzure_ApplyTemplate (0.01s)
    --- PASS: TestAzure_ApplyTemplate/basic_stack (0.00s)
    --- PASS: TestAzure_ApplyTemplate/custom_stack (0.00s)
```

ppknap replied:
Should be good for now. However, this method does not work when shadowed element is inside slice.

rjeczalik committed:
provider: finish inline doc

rjeczalik commented:
What about we merge azure before deploy? The feature would be a ~nop without client support.

gokmen commented:
@cihangir we can merge this since it seems it won't affect deploy afais.
